service-base/src/main/java/org/eclipse/hono/service/metric/DeviceConnectionTimeTracker.java

```diff
@@ -0,0 +1,164 @@
+
+    private static final long DEVICE_CONNECTION_TIME_RECORDING_INTERVAL_IN_MS = TimeUnit.SECONDS.toMillis(10);
+    protected final Logger log = LoggerFactory.getLogger(DeviceConnectionTimeTracker.class);
+    private final AtomicLong noOfDeviceConnections;
+    private final Consumer<Long> recorder;
+    private final AtomicReference<Instant> startInstant = new AtomicReference<>();
+    private final String tenantId;
+    private final long timerId;
+    private final Vertx vertx;
+
+    private DeviceConnectionTimeTracker(
+            final String tenantId, final Vertx vertx,
+            final long noOfDeviceConnections,
+            final Consumer<Long> recorder) {
+        this.tenantId = tenantId;
+        this.noOfDeviceConnections = new AtomicLong(noOfDeviceConnections);
+        this.recorder = recorder;
+        this.startInstant.set(Instant.now());
+        this.vertx = vertx;
+        this.timerId = this.vertx.setPeriodic(DEVICE_CONNECTION_TIME_RECORDING_INTERVAL_IN_MS, id -> report());
+        log.trace("Stared a device connection time tracker for the tenant [{}].", tenantId);
+    }
+
+    /**
+     * Reports the device connection time and then updates the number of device connections with the given value.
+     * 
+     * @param noOfDeviceConnections The number of device connections.
+     * @return a reference to this for fluent use.
+     */
+    public DeviceConnectionTimeTracker updateNoOfDeviceConnections(final long noOfDeviceConnections) {
+        report();
+        this.noOfDeviceConnections.set(noOfDeviceConnections);
+        log.trace("Updated number of device connections for the tenant [{}] to [{}]", tenantId,
+                noOfDeviceConnections);
+        if (noOfDeviceConnections == 0) {
+            vertx.cancelTimer(timerId);
+            log.trace("Stopped the device connection time tracker for the tenant [{}].", tenantId);
+            return null;
+        }
+        return this;
+    }
+
+    private void report() {
+        final long deviceConnectionTime = noOfDeviceConnections.get()
+                * Duration.between(startInstant.getAndSet(Instant.now()), Instant.now()).toMillis();
+        recorder.accept(deviceConnectionTime);
+        log.trace(
+                "Reported device connection time [tenant : {}, noOfDeviceConnections: {}, totalConnectionTime: {}].",
+                tenantId, noOfDeviceConnections.get(), deviceConnectionTime);
+    }
+
+    /**
+     * A builder class for {@link DeviceConnectionTimeTracker}.
+     */
+    public static final class Builder {
+
+        private long noOfDeviceConnections;
+        private Vertx vertx;
+        private Consumer<Long> recorder;
+        private String tenantId;
+
+        private Builder() {
+            // prevent instantiation
+        }
+
+        /**
+         * Creates an instance of the Builder.
+         * 
+         * @return a reference to this for fluent use.
+         */
+        public static Builder createInstance() {
```

Comment:
Apart from creating an instance of the Builder, it has no other use. This can be removed as now `forTenant(...)` is a static factory method and returns an instance of the Builder.